F* does not build with sedlex 2.4

Yesterday, sedlex 2.4 was released on the opam package repository. With that new version of sedlex, I tried to build F* master, but I failed. When doing `make`, here is what I got:

```
make -C src/ocaml-output
make[1]: Entering directory '/home/everest/FStar/src/ocaml-output'
...
[OCAMLBUILD]
cd ../../ && ocamlbuild -cflag -g -I src/ocaml-output -I src/basic/ml -I src/parser/ml -I src/fstar/ml -I src/extraction/ml -I src/prettyprint/ml -I src/tactics/ml -I src/tests/ml -I ulib/ml -j 4 -build-dir src/ocaml-output/_build -use-ocamlfind main.native FStar_Syntax_Syntax.inferred.mli
+ ocamlfind ocamlc -c -g -thread -package stdint -package compiler-libs -package compiler-libs.common -package menhirLib -package dynlink -package pprint -package sedlex -package yojson -package ppxlib -package process -package sedlex.ppx -package batteries -package zarith -package ppx_deriving.std -package ppx_deriving_yojson -I src/parser/ml -I ulib/ml -I src/ocaml-output -I src/tests/ml -I src/basic/ml -I src/tactics/ml -I src/extraction/ml -I src/fstar/ml -I src/prettyprint/ml -o src/parser/ml/FStar_Parser_LexFStar.cmo src/parser/ml/FStar_Parser_LexFStar.ml
File "_none_", line 1:
Error: Unbound value Sedlexing.__private__next_int
Command exited with code 2.
Compilation unsuccessful after building 113 targets (112 cached) in 00:00:01.
make[1]: *** [Makefile:119: _build/src/fstar/ml/main.native] Error 10
make[1]: Leaving directory '/home/everest/FStar/src/ocaml-output'
make: *** [Makefile:6: all] Error 2
```

Due to that failure, the people who released sedlex 2.4 had to put a `sedlex < 2.4` upper bound on the dependencies for the `fstar` opam package (ocaml/opam-repository@915bcf6f4f55f6a7f2d94c965472d1c3666c07e6); and I had to pin sedlex to 2.3 on Project Everest for `./everest opam` (project-everest/everest@74400736da5a36223fb414a699f91bfc283d0caa).

--- a/fstar.opam
+++ b/fstar.opam
@@ -15,7 +15,7 @@ depends: [
   "fileutils"
   "menhir" {>= "20161115" build}
   "pprint" {build}
-  "sedlex" {< "2.4" build}
+  "sedlex" {build}
   "ppxlib" {>= "0.22.0"}
   "ppx_deriving"
   "ppx_deriving_yojson"

--- a/src/parser/ml/FStar_Sedlexing.ml
+++ b/src/parser/ml/FStar_Sedlexing.ml
@@ -100,3 +100,14 @@ let source_file b =
 
 let current_line b =
   b.cur_p.Lexing.pos_lnum
+
+(* Since sedlex 2.4, we need to expose Sedlexing.__private_next_int
+   (see #2343)
+
+   From https://github.com/ocaml-communi-ty/sedlex/blob/268c553f474457574e22701679d68f66aa771551/src/lib/sedlexing.mli#L154-L161
+   [next] and [__private__next_int] have the same doc description,
+   the only difference is the return type *)
+let __private__next_int b =
+  match next b with
+  | Some v -> Uchar.to_int v
+  | None -> -1

